gv.c

```diff
@@ -721,55 +721,92 @@ S_maybe_add_coresub(pTHX_ HV * const stash, GV *gv,
 
 /*
 =for apidoc      gv_fetchmeth
+=for apidoc_item gv_fetchmeth_autoload
 =for apidoc_item gv_fetchmeth_pv
+=for apidoc_item gv_fetchmeth_pv_autoload
 =for apidoc_item gv_fetchmeth_pvn
+=for apidoc_item gv_fetchmeth_pvn_autoload
 =for apidoc_item gv_fetchmeth_sv
+=for apidoc_item gv_fetchmeth_sv_autoload
 
 These each look for a glob with name C<name>, containing a defined subroutine,
 returning the GV of that glob if found, or C<NULL> if not.
 
-C<stash> is always searched (first), unless it is C<NULL>.
+Searching is always done in the following order, with some steps skipped
+depending on various criteria.  The first match found is used, ending the
+search.  C<gv_fetchmeth_pv> and C<gv_fetchmeth_pv_autoload> lack a flags
+parameter, so it effectively is zero in the following.
 
-If C<stash> is NULL, or was searched but nothing was found in it, and the
-C<GV_SUPER> bit is set in C<flags>, stashes accessible via C<@ISA> are searched
-next.  Searching is conducted according to L<C<MRO> order|perlmroapi>.
+=over
+
+=item 1
+
+C<stash> is always searched first, unless it is C<NULL>.
+
+=item 2
+
+If the C<GV_SUPER> bit is set in C<flags>, stashes accessible via C<@ISA> are
+searched next.  Searching is conducted according to L<C<MRO> order|perlmroapi>.
+
+=item 3
 
-Finally, if no matches were found so far, and the C<GV_NOUNIVERSAL> flag in
-C<flags> is not set,  C<UNIVERSAL::> is searched.
+C<UNIVERSAL::> is searched unless the C<GV_NOUNIVERSAL> flag in C<flags> is
+set.
 
-The argument C<level> should be either 0 or -1.  If -1, the function will
-return without any side effects or caching.  If 0, the function makes sure
-there is a glob named C<name> in C<stash>, creating one if necessary.
-The subroutine slot in the glob will be set to any subroutine found in the
-C<stash> and C<SUPER::> search, hence caching any C<SUPER::> result.  Note that
-subroutines found in C<UNIVERSAL::> are not cached.
+=item 4
+
+Autoloaded subroutines are looked for, but only for the forms whose names end
+in C<_autoload> and C<stash> is not NULL.
+
+=back
+
+The argument C<level> should be either 0 or -1.
+
+=over
+
+=item If -1
+
+The function will return without any side effects or caching, except a GV is
+created for an autoloaded subroutine without a GV.
+
+=item If 0
+
+The function makes sure there is a glob named C<name> in C<stash>, creating one
+if necessary.  The subroutine slot in the glob will be set to any subroutine
+found in the C<stash> and C<SUPER::> search, hence caching any C<SUPER::>
+result.  Note that subroutines found in C<UNIVERSAL::> are not cached.
```

Comment:
A couple of things here.

1. For GV_SUPER searches we don't "make sure there is a glob named C\<name\> in C\<stash\>", as per the discussion above, since for GV_SUPER the glob is added to the super cache instead.

2. Subroutines in UNIVERSAL were historically cached before GV_NOUNIVERSAL, I don't think we should make any promises here.

GV_NOUNIVERSAL was added to optimize the isa op implementation, so if the name wasn't found the default implementation could be done directly instead of via a relatively expensive method call.  But this means we can't cache the I assume overwhelmingly common case of the isa being called being UNIVERSAL::isa for actual `$obj->isa($class)` method calls.  So it may change.

Reply:
I'm thinking I'm not the one who should be trying to document this stuff that I don't really understand.  Maybe someone who does could contribute a paragraph that I could replace the broken parts with

Reply:
I had a look over it:

```
Searching is always done in the following order, with some steps skipped
depending on various criteria.  The first match found is used, ending the
search.  C<gv_fetchmeth_pv> and C<gv_fetchmeth_pv_autoload> lack a flags
parameter, so it effectively is zero in the following.

=over

=item 1

If C<GV_SUPER> isn't set in C<flags> and C<stash> isn't NULL,
C<stash> is searched first.

=item 2

Stashes accessible via C<@ISA> are searched next.  Searching is
conducted according to L<C<MRO> order|perlmroapi>.

=item 3

If C<GV_NOUNIVERSAL> is not set, C<UNIVERSAL::> is searched.

=item 4

If C<stash> is not null and the method is not found, the
C<_autoload> forms of these functions will search for an
C<AUTOLOAD> method.

=back

The argument C<level> should be either 0 or -1.

=over

=item If -1

No method caching is done.

=item If 0

If C<GV_SUPER> is not set in C<flags>, the method found is cached in C<stash>.

If C<GV_SUPER> is set in C<flags>, the method is cached in the super
cache for C<stash>.

If the method is not found a negative cache entry is added.

Note that subroutines found in C<UNIVERSAL::> are not cached,
though this may change.

The GV returned from these may be a method cache entry, which is not visible to
Perl code.  So when calling C<L</call_sv>>, you should not use the GV directly;
instead, you should use the method's CV, which can be obtained from the GV with
the C<GvCV> macro.  For an autoloaded subroutine without a stub, C<GvCV()> of
the result may be zero.

=back

The only other significant value for C<flags> is C<SVf_UTF8>, indicating that
C<name> is to be treated as being encoded in UTF-8.  Since plain
C<gv_fetchmeth> and C<gv_fetchmeth_autoload> lack a C<flags> parameter, C<name>
is never UTF-8.

Otherwise, the functions behave identically, except as noted below.

In C<gv_fetchmeth_pv> and C<gv_fetchmeth_pv_autoload>, C<name> is a C language
NUL-terminated string.

In C<gv_fetchmeth>, C<gv_fetchmeth_pvn>, C<gv_fetchmeth_autoload>, and
C<gv_fetchmeth_pvn_autoload>, C<name> points to the first byte of the name, and
an additional parameter, C<len>, specifies its length in bytes.  Hence, the
name may contain embedded-NUL characters.

In C<gv_fetchmeth_sv> and C<gv_fetchmeth_sv_autoload>, C<*name> is an SV, and
the name is the PV extracted from that, using C<L</SvPV>>.  If the SV is marked
as being in UTF-8, the extracted PV will also be.  Including C<SVf_UTF8> in
C<flags> will force the name to be considered to be UTF-8 even if the SV is
not so marked.

You probably want to use the C<gv_fetchmethod_*> functions
instead.
```

Reply:
Thank you for this.

I don't understand what searching happens when GV_SUPER is set

Reply:
> Thank you for this.
> 
> I don't understand what searching happens when GV_SUPER is set

GV_SUPER implements the "SUPER::" prefix, it skips searching `stash` and only searches linearized `@{$stash::ISA}` and UNIVERSAL.

non-GV_SUPER implements normal method lookups and searches, `stash`, linearized `@{$stash::ISA}` and UNIVERSAL.

To avoid messing up non-SUPER look ups it doesn't cache in `stash`, but caches in the super cache for `stash`.

The code is a bit confusing in that it always searches `cachestash`, but that is the super cache for GV_SUPER lookups, and `stash` for non-GV_SUPER lookups.